comptime: struct fields as maps with different key/value types can't be deduced

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.12 3d7be5f, press to see full `v doctor` output</summary>

|V full version      |V 0.4.12 c1d1efb2091e989dbb09e38b308c82c2cade1aa8.3d7be5f
|:-------------------|:-------------------
|OS                  |linux, Ubuntu 22.04.5 LTS
|Processor           |4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-4150 CPU @ 3.50GHz
|Memory              |0.66GB/3.7GB
|                    |
|V executable        |/home/jorge/v/v
|V last modified time|2025-11-14 18:18:14
|                    |
|V home dir          |OK, value: /home/jorge/v
|VMODULES            |OK, value: /home/jorge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/jorge/bugs
|                    |
|Git version         |git version 2.34.1
|V git status        |weekly.2025.16-1062-g3d7be5f5-dirty
|.git/config present |true
|                    |
|cc version          |cc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0
|gcc version         |gcc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0
|clang version       |Ubuntu clang version 14.0.0-1ubuntu1.1
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.5 ()
|glibc version       |ldd (Ubuntu GLIBC 2.35-0ubuntu3.11) 2.35

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  251114.v && 251114`
```v
struct Maps {
	a map[string]int
	b map[string]int
	c map[string]f64
}

fn decode_struct[T](t T) {
	$for f in T.fields {
		$if f is $map {
			k, v := map_key_value(t.$(f.name))
			dump('${typeof(k).name} ${typeof(v).name}')
		}
	}
}

fn map_key_value[K,V](m map[K]V) (K, V) {
	return K{}, V{}
}

fn main() {
	decode_struct(Maps{})
}


```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/251114.01KA1SHHGVS38JTHES0QR827FJ.tmp.c:6094: error: '{' expected (got ";")
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

get key and value types from different struct map fields.

**Details**

```
/tmp/v_1000/251114.01KA1T1Q1RFMQF6YHSY2RWHCD0.tmp.c:6094: error: '{' expected (got ";")
=================================================================================
======== Output of the C Compiler (cc) ========
/tmp/v_1000/251114.01KA1T1Q1RFMQF6YHSY2RWHCD0.tmp.c: In function ‘main__decode_struct_T_main__Maps’:
/tmp/v_1000/251114.01KA1T1Q1RFMQF6YHSY2RWHCD0.tmp.c:6094:66: error: invalid initializer
 6094 |                                 multi_return_string_f64 mr_146 = main__map_key_value_T_string_int(t.a);
      |                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_1000/251114.01KA1T1Q1RFMQF6YHSY2RWHCD0.tmp.c:6121:66: error: invalid initializer
 6121 |                                 multi_return_string_f64 mr_146 = main__map_key_value_T_string_int(t.b);
      |                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
===============================================
```

I want to check the key and value types of all struct's maps fields.
The compiler works when *all* fields maps are of *both* same key and same value type so:
```v
struct Maps {
	a map[string]int
	b map[string]int
//	c map[string]f64
}
```
works producing:
```
[251114.v:11] '${typeof(k).name} ${typeof(v).name}': string int
[251114.v:11] '${typeof(k).name} ${typeof(v).name}': string int
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -1112,17 +1112,51 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 fn (mut g Gen) gen_multi_return_assign(node &ast.AssignStmt, return_type ast.Type, return_sym ast.TypeSymbol) {
 	// multi return
 	// TODO: Handle in if_expr
-	mr_var_name := 'mr_${node.pos.pos}'
-	mut is_option := return_type.has_flag(.option)
-	mut mr_styp := g.styp(return_type.clear_flag(.result))
+	mut ret_type := return_type
+	mut ret_sym := return_sym
+	mut suffix := ''
+	if g.comptime.inside_comptime_for && node.right[0] is ast.CallExpr {
+		call_expr := node.right[0] as ast.CallExpr
+		if call_expr.concrete_types.len > 0 && return_sym.info is ast.MultiReturn
+			&& g.comptime.comptime_for_field_var != '' {
+			field_type := g.comptime.comptime_for_field_type
+			field_sym := g.table.sym(field_type)
+			if field_sym.info is ast.Map {
+				map_info := field_sym.info as ast.Map
+				ret_type = g.table.find_or_register_multi_return([map_info.key_type, map_info.value_type])
+				ret_sym = *g.table.sym(ret_type)
+			}
+		}
+		if g.comptime.comptime_for_field_var != '' {
+			suffix = '_${g.comptime.comptime_for_field_value.name}'
+		}
+	}
+	mr_var_name := 'mr_${node.pos.pos}${suffix}'
+	mut is_option := ret_type.has_flag(.option)
+	mut mr_styp := g.styp(ret_type.clear_flag(.result))
 	if node.right[0] is ast.CallExpr && node.right[0].or_block.kind != .absent {
 		is_option = false
-		mr_styp = g.styp(return_type.clear_option_and_result())
+		mr_styp = g.styp(ret_type.clear_option_and_result())
 	}
 	g.write('${mr_styp} ${mr_var_name} = ')
 	g.expr(node.right[0])
 	g.writeln(';')
-	mr_types := (return_sym.info as ast.MultiReturn).types
+	mr_types := (ret_sym.info as ast.MultiReturn).types
+	mut recompute_types := false
+	if g.comptime.inside_comptime_for && node.right[0] is ast.CallExpr {
+		call_expr := node.right[0] as ast.CallExpr
+		if call_expr.concrete_types.len > 0 && g.comptime.comptime_for_field_var != ''
+			&& return_sym.info is ast.MultiReturn {
+			recompute_types = true
+			for i, mut lx in &node.left {
+				if mut lx is ast.Ident && lx.kind != .blank_ident {
+					if mut lx.obj is ast.Var {
+						lx.obj.typ = mr_types[i]
+					}
+				}
+			}
+		}
+	}
 	for i, lx in node.left {
 		mut cur_indexexpr := -1
 		mut is_auto_heap := false
@@ -1141,7 +1175,8 @@ fn (mut g Gen) gen_multi_return_assign(node &ast.AssignStmt, return_type ast.Typ
 		if lx is ast.IndexExpr && g.cur_indexexpr.len > 0 {
 			cur_indexexpr = g.cur_indexexpr.index(lx.pos.pos)
 		}
-		styp := if ident.name in g.defer_vars { '' } else { g.styp(node.left_types[i]) }
+		left_type := if recompute_types { mr_types[i] } else { node.left_types[i] }
+		styp := if ident.name in g.defer_vars { '' } else { g.styp(left_type) }
 		if node.op == .decl_assign {
 			g.write('${styp} ')
 		}
@@ -1150,14 +1185,14 @@ fn (mut g Gen) gen_multi_return_assign(node &ast.AssignStmt, return_type ast.Typ
 		}
 		noscan := if is_auto_heap { g.check_noscan(return_type) } else { '' }
 		mut aligned := 0
-		sym := g.table.final_sym(node.left_types[i])
+		sym := g.table.final_sym(left_type)
 		if sym.info is ast.Struct {
 			if attr := sym.info.attrs.find_first('aligned') {
 				aligned = if attr.arg == '' { 0 } else { attr.arg.int() }
 			}
 		}
-		if node.left_types[i].has_flag(.option) {
-			base_typ := g.base_type(node.left_types[i])
+		if left_type.has_flag(.option) {
+			base_typ := g.base_type(left_type)
 			tmp_var := if is_auto_heap {
 				if aligned != 0 {
 					'HEAP_align(${styp}, ${mr_var_name}.arg${i}, ${aligned})'

--- a/vlib/v/gen/c/str_intp.v
+++ b/vlib/v/gen/c/str_intp.v
@@ -191,6 +191,11 @@ fn (mut g Gen) str_val(node ast.StringInterLiteral, i int, fmts []u8) {
 	fmt := fmts[i]
 	typ := g.unwrap_generic(node.expr_types[i])
 	typ_sym := g.table.sym(typ)
+	if g.comptime.inside_comptime_for && expr is ast.SelectorExpr && expr.field_name == 'name'
+		&& expr.expr is ast.TypeOf {
+		g.expr(expr)
+		return
+	}
 	if typ == ast.string_type && g.comptime.comptime_for_method == unsafe { nil } {
 		if g.inside_vweb_tmpl {
 			g.write('${g.vweb_filter_fn_name}(')
@@ -278,6 +283,24 @@ fn (mut g Gen) string_inter_literal(node ast.StringInterLiteral) {
 	mut node_ := unsafe { node }
 	mut fmts := node_.fmts.clone()
 	for i, mut expr in node_.exprs {
+		mut type_to_use := node_.expr_types[i]
+		if g.comptime.inside_comptime_for && mut expr is ast.SelectorExpr {
+			if expr.expr is ast.TypeOf && expr.field_name == 'name' {
+				// This is typeof(var).name
+				typeof_expr := expr.expr as ast.TypeOf
+				if typeof_expr.expr is ast.Ident {
+					ident_name := (typeof_expr.expr as ast.Ident).name
+					// Check if this ident might be from a multi-return in the current scope
+					// For now, force re-evaluation by looking it up in the comptime context
+					if obj := typeof_expr.expr.scope.find(ident_name) {
+						if obj is ast.Var {
+							// Use the var's actual type, which might be correct in codegen context
+							type_to_use = obj.typ
+						}
+					}
+				}
+			}
+		}
 		if g.comptime.is_comptime(expr) {
 			ctyp := g.type_resolver.get_type_or_default(expr, node_.expr_types[i])
 			if ctyp != ast.void_type {
@@ -292,6 +315,8 @@ fn (mut g Gen) string_inter_literal(node ast.StringInterLiteral) {
 					fmts[i] = g.get_default_fmt(ctyp, typ)
 				}
 			}
+		} else {
+			node_.expr_types[i] = type_to_use
 		}
 	}
 	g.write2('builtin__str_intp(', node.vals.len.str())

--- a/vlib/v/tests/comptime/comptime_map_fields_decode_test.v
+++ b/vlib/v/tests/comptime/comptime_map_fields_decode_test.v
@@ -1,0 +1,24 @@
+struct Maps {
+	a map[string]int
+	b map[string]bool
+	c map[string]f64
+}
+
+fn decode_struct[T](t T) []string {
+	mut decoded := []string{}
+	$for f in T.fields {
+		$if f is $map {
+			k, v := map_key_value(t.$(f.name))
+			decoded << '${typeof(k).name} ${typeof(v).name}'
+		}
+	}
+	return decoded
+}
+
+fn map_key_value[K, V](m map[K]V) (K, V) {
+	return K{}, V{}
+}
+
+fn test_main() {
+	assert decode_struct(Maps{}) == ['string int', 'string bool', 'string f64']
+}

